### DISCUSSION TITLE
Fix: use sigmoid function for reranking models to return scores from 0 to 1.

### DISCUSSION THREAD

**Discussion by BarerRocket2678** [open-webui/open-webui#19999] [General]
This fix makes Reranking models return a score of 0 to 1 instead of an arbitrary value. (like -10 to 10 in the case of cross encoder Rerankers) I have the pull request ready but not yet created. Attached is the file that has been modified. 
[retrieval.py](https://github.com/user-attachments/files/24202642/retrieval.py)

Changes:
Added `import torch` on line 7
Added `activation_fn=torch.nn.Sigmoid(),` to `sentence_transformers.CrossEncoder(...)` on line 191
Removed `import torch` (formerly) on line 314
Removed `import torch` (formerly) on line 895

Reranking models that were tested and work with this change:
- cross-encoder/ms-marco-TinyBERT-L2-v2
- cross-encoder/ms-marco-MiniLM-L2-v2
- cross-encoder/ms-marco-MiniLM-L4-v2
- cross-encoder/ms-marco-MiniLM-L6-v2
- cross-encoder/ms-marco-MiniLM-L12-v2
- cross-encoder/ms-marco-electra-base
- BAAI/bge-reranker-v2-m3

**Comment by mahenning** [comment]
Please create a pull request (PR) if you have code fixes. A step by step guide for that is here: https://docs.github.com/articles/creating-a-pull-request

You should not import torch there just like that. As you can see in the other two torch imports that you removed, torch is only imported situational, e.g. when `DEVICE_TYPE == "cuda"`. Thats because torch is not always required. For example, if OWUI is configured with only external embeding/reranker, the code for loading a `CrossEncoder` is never called and the torch import is not used. So you should put the torch import not at the top, but e.g. there:
```python
# line ~180
            else:
                import sentence_transformers
                import torch  # <--- this would be a good place
                
                try:
                    rf = sentence_transformers.CrossEncoder(
                        get_model_path(reranking_model, auto_update),
                        device=DEVICE_TYPE,
                        trust_remote_code=RAG_RERANKING_MODEL_TRUST_REMOTE_CODE,
                        backend=SENTENCE_TRANSFORMERS_CROSS_ENCODER_BACKEND,
                        model_kwargs=SENTENCE_TRANSFORMERS_CROSS_ENCODER_MODEL_KWARGS,
                    )
```

I agree that Sigmoid should be default. In all other models I tested it was defaulted to, but in the case of the `cross-encoder/ms-marco-*` models, they have [a line in their configs](https://huggingface.co/cross-encoder/ms-marco-MiniLM-L6-v2/blob/main/config.json#L30): 
> "sbert_ce_default_activation_function": "torch.nn.modules.linear.Identity"

In that case, the fallback activation function is indeed the identity function.

**Comment by jfahrenkrug** [comment]
@BarerRocket2678 This is a great proposal. Was a PR for this ever created/merged? 
Update: Nevermind, found it: https://github.com/open-webui/open-webui/pull/20020 Thank you!

**Reply by Classic298** [reply]
it was merged later yes
